### PR TITLE
Add unsafe memory option for modded games

### PR DESCRIPTION
**Pull request by EwanMac3**
Adds a setting that ignores memory permission checks, allowing certain homebrew and mod menus to function within Ryujinx like on actual hardware.

### CODE REVIEW

**Comment by MutantAura** [comment]
You may want to make these changes without formatting (or allowing your IDE to auto-format) every file that was changed. Lots of structure has been changed that aren't relevant to the PR changes.

File: Ryujinx.HLE/HLEConfiguration.cs [hunk 1]
```diff
@@ -153,29 +158,13 @@ public class HLEConfiguration
...
-                                bool                   enableInternetAccess,
-                                IntegrityCheckLevel    fsIntegrityCheckLevel,
-                                int                    fsGlobalAccessLogMode,
-                                long                   systemTimeOffset,
-                                string                 timeZone,
-                                MemoryManagerMode      memoryManagerMode,
-                                bool                   ignoreMissingServices,
-                                AspectRatio            aspectRatio,
-                                float                  audioVolume)
+        public HLEConfiguration(VirtualFileSystem virtualFileSystem, LibHacHorizonManager libHacHorizonManager,
```
Comment (gdkchan) [inline Ryujinx.HLE/HLEConfiguration.cs:161]: Why was this completely reformatted?

File: Ryujinx.HLE/HOS/Kernel/SupervisorCall/Syscall.cs [hunk 2]
```diff
@@ -8,13 +8,16 @@
 using Ryujinx.HLE.HOS.Kernel.Process;
 using Ryujinx.HLE.HOS.Kernel.Threading;
 using System;
+using System.Management;
 using System.Threading;
 
 namespace Ryujinx.HLE.HOS.Kernel.SupervisorCall
 {
-    [SvcImpl]
+[SvcImpl]
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Kernel/SupervisorCall/Syscall.cs:16]: Why was this changed.

File: Ryujinx.HLE/HOS/Kernel/SupervisorCall/Syscall.cs [hunk 3]
```diff
@@ -8,13 +8,16 @@
...
     class Syscall
     {
+        public static bool allowUnsafeMemAccess;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Kernel/SupervisorCall/Syscall.cs:19]: This should not be using a static field. And it's not following our naming conventions. No changes on this file are necessary to do what you want, you can just force `AllowCodeMemoryForJit` on `KProcess` to true based on the setting.

File: Ryujinx.Headless.SDL2/Program.cs [hunk 4]
```diff
@@ -484,7 +484,7 @@ private static Switch InitializeEmulationContext(WindowBase window, Options opti
                                                                   options.MemoryManagerMode,
                                                                   (bool)options.IgnoreMissingServices,
                                                                   options.AspectRatio,
-                                                                  options.AudioVolume);
+                                                                  options.AudioVolume, options.UnsafeMem);
```
Comment (gdkchan) [inline Ryujinx.Headless.SDL2/Program.cs:487]: `options.UnsafeMem` should be on a new line like the other parameters.

File: Ryujinx/Ui/MainWindow.cs [hunk 5]
```diff
@@ -574,7 +574,7 @@ private void InitializeSwitchInstance()
                                                                           ConfigurationState.Instance.System.MemoryManagerMode,
                                                                           ConfigurationState.Instance.System.IgnoreMissingServices,
                                                                           ConfigurationState.Instance.Graphics.AspectRatio,
-                                                                          ConfigurationState.Instance.System.AudioVolume);
+                                                                          ConfigurationState.Instance.System.AudioVolume, ConfigurationState.Instance.System.UnsafeMem.Value);
```
Comment (gdkchan) [inline Ryujinx/Ui/MainWindow.cs:577]: Should be on new line too.

File: Ryujinx/Ui/Windows/SettingsWindow.glade [hunk 6]
```diff
@@ -1,2696 +1,2714 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!-- Generated with glade 3.38.2 -->
 <interface>
-  <requires lib="gtk+" version="3.20"/>
```
Comment (gdkchan) [inline Ryujinx/Ui/Windows/SettingsWindow.glade:4]: Please keep the indentation on the file as before, no need to completely re-format it.

**Review by gdkchan — COMMENTED** [review]
I don't think `UnsafeMem` is a good name for this feature because it does not correctly describe what it actually does. Maybe something like `Enable code memory kernel patches` and `EnableCodeKernelPatches`.

The problem with exposing something like that is that people will enable it and break some mods. And then they will ask in support or open a issue here saying that it doesn't work.

Right now we already enable it for homebrew, so it should be only a problem for mods. I think if we made the option toggle-able, would make more sense to just enable or disable it for everything.

File: Ryujinx/Ui/Windows/SettingsWindow.glade (hunk 6 above)
Comment (EwanMac3) [reply Ryujinx/Ui/Windows/SettingsWindow.glade:4]: This one was done by my IDE, I didn't notice until it was already pointed out. Thanks for the tips, I'll revise it.

File: Ryujinx.HLE/HOS/Kernel/Process/KProcess.cs [hunk 7]
```diff
@@ -103,7 +103,7 @@ public KProcess(KernelContext context, bool allowCodeMemoryForJit = false) : bas
 
             Capabilities = new KProcessCapabilities();
 
-            AllowCodeMemoryForJit = allowCodeMemoryForJit;
+            AllowCodeMemoryForJit = context.Device.Configuration.AllowJitCodeMem;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Kernel/Process/KProcess.cs:106]: You should pass the setting value on the `allowCodeMemoryForJit` parameter instead, then you don't need to make any change on this file.

File: Ryujinx.HLE/HLEConfiguration.cs [hunk 8]
```diff
@@ -175,7 +180,8 @@ public class HLEConfiguration
                                 MemoryManagerMode      memoryManagerMode,
                                 bool                   ignoreMissingServices,
                                 AspectRatio            aspectRatio,
-                                float                  audioVolume)
+                                float                  audioVolume,
+                                bool                   AllowJitCodeMemOpt)
```
Comment (gdkchan) [inline Ryujinx.HLE/HLEConfiguration.cs:184]:
```suggestion
                                bool                   allowJitCodeMem)
```

File: Ryujinx.HLE/HLEConfiguration.cs [hunk 9]
```diff
@@ -200,6 +206,7 @@ public class HLEConfiguration
             IgnoreMissingServices  = ignoreMissingServices;
             AspectRatio            = aspectRatio;
             AudioVolume            = audioVolume;
+            AllowJitCodeMem        = AllowJitCodeMemOpt;
```
Comment (gdkchan) [inline Ryujinx.HLE/HLEConfiguration.cs:209]:
```suggestion
            AllowJitCodeMem        = allowJitCodeMem;
```

Comment (gdkchan) [reply Ryujinx.HLE/HLEConfiguration.cs:209]: I'd rename `AllowJitCodeMem` to `AllowJitCodeMemory` though. Same for the lower case variant.

File: Ryujinx.Headless.SDL2/Options.cs [hunk 10]
```diff
@@ -176,5 +176,8 @@ public class Options
 
         [Value(0, MetaName = "input", HelpText = "Input to load.", Required = true)]
         public string InputPath { get; set; }
+
+        [Option("allow-jit-code-mem", Required = false, Default = false, HelpText = "Allows mods to access areas of memory they normally shouldn't. Don't enable this setting unless a mod is crashing because it needs it.")]
```
Comment (gdkchan) [inline Ryujinx.Headless.SDL2/Options.cs:180]: I think the setting needs a better description. Also a mod causing crashes could be due to several reasons, I don't think just enabling this option if they crash is a good advice.

Maybe "Enable kernel code memory patches. This is required by some mods and homebrew that needs to modify code at runtime. Leave it disabled if unsure."

**Comment by riperiperi** [comment]
I'm not sure a configuration option is needed for this at all. If the assumption is that any homebrew or modded games can only be run when the OS is behaving like atmosphere, then wouldn't it be enough to automatically enable this when mods are enabled or homebrew is launched?

**Comment by marysaka** [comment]
Closing as stale.
